File: src/pipeline/normalizer.ts
import { createHash } from 'crypto';
import { RawListing } from '../scrapers/base';

export interface NormalizedListing {
    sourceListingId: string;
    sourceUrl: string;
    title: string;
    description: string | null;
    descriptionClean: string | null;
    askingPrice: bigint | null;
    askingPriceRaw: string | null;
    revenue: bigint | null;
    revenueRaw: string | null;
    cashFlow: bigint | null;
    cashFlowRaw: string | null;
    cashFlowType: string | null;
    multiple: number | null;
    industry: string | null;
    industryNormalized: string | null;
    businessType: string | null;
    city: string | null;
    state: string | null;
    stateCode: string | null;
    country: string;
    zipCode: string | null;
    brokerName: string | null;
    brokerCompany: string | null;
    brokerPhone: string | null;
    brokerEmail: string | null;
    yearEstablished: number | null;
    employees: number | null;
    isFranchise: boolean | null;
    sellerFinancing: boolean | null;
    isAbsenteeOwner: boolean | null;
    isHomeBased: boolean | null;
    hasRealEstate: boolean | null;
    reasonForSelling: string | null;
    dedupeHash: string;
}

export function normalizeListing(raw: RawListing): NormalizedListing {
    const askingPrice = parseCurrency(raw.askingPriceRaw);
    const cashFlow = parseCurrency(raw.cashFlowRaw);
    const revenue = parseCurrency(raw.revenueRaw);
    const stateCode = normalizeState(raw.state);

    return {
        sourceListingId: raw.sourceListingId,
        sourceUrl: raw.sourceUrl,
        title: cleanText(raw.title),
        description: raw.description || null,
        descriptionClean: raw.description ? stripHtml(raw.description) : null,
        askingPrice,
        askingPriceRaw: raw.askingPriceRaw || null,
        revenue,
        revenueRaw: raw.revenueRaw || null,
        cashFlow,
        cashFlowRaw: raw.cashFlowRaw || null,
        cashFlowType: raw.cashFlowType || null,
        multiple:
            askingPrice && cashFlow && cashFlow > 0n
                ? Number(askingPrice) / Number(cashFlow)
                : null,
        industry: raw.industry || null,
        industryNormalized: mapIndustry(raw.industry, raw.title, raw.description),
        businessType: raw.businessType || null,
        city: raw.city ? titleCase(raw.city.slice(0, 100)) : null,
        state: stateCode ? STATE_NAMES[stateCode] : (raw.state || null),
        stateCode,
        country: 'US',
        zipCode: raw.zipCode || null,
        brokerName: raw.brokerName || null,
        brokerCompany: raw.brokerCompany || null,
        brokerPhone: normalizePhone(raw.brokerPhone),
        brokerEmail: raw.brokerEmail?.toLowerCase() || null,
        yearEstablished: raw.yearEstablished || null,
        employees: raw.employees || null,
        isFranchise: raw.isFranchise ?? null,
        sellerFinancing: raw.sellerFinancing ?? null,
        isAbsenteeOwner: raw.isAbsenteeOwner ?? null,
        isHomeBased: raw.isHomeBased ?? null,
        hasRealEstate: raw.hasRealEstate ?? null,
        reasonForSelling: raw.reasonForSelling || null,
        dedupeHash: generateDedupeHash(raw),
    };
}

// ---- Currency Parsing ----

function parseCurrency(raw?: string): bigint | null {
    if (!raw) return null;
    let cleaned = raw.replace(/[^0-9.,kKmMbB-]/g, '').trim();
    if (!cleaned || cleaned === '-') return null;

    let multiplier = 100; // cents
    const suffix = cleaned.slice(-1).toUpperCase();
    if (suffix === 'K') { multiplier = 100_000; cleaned = cleaned.slice(0, -1); }
    else if (suffix === 'M') { multiplier = 100_000_000; cleaned = cleaned.slice(0, -1); }
    else if (suffix === 'B') { multiplier = 100_000_000_000; cleaned = cleaned.slice(0, -1); }

    cleaned = cleaned.replace(/,/g, '');
    const value = parseFloat(cleaned);
    if (isNaN(value) || value <= 0) return null;
    return BigInt(Math.round(value * multiplier));
}

// ---- Text Cleaning ----

function cleanText(text: string): string {
    return text.replace(/\s+/g, ' ').trim().slice(0, 500);
}

function stripHtml(html: string): string {
    return html
        .replace(/<[^>]+>/g, ' ')
        .replace(/&nbsp;/g, ' ')
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/\s+/g, ' ')
        .trim();
}

function titleCase(str: string): string {
    return str.toLowerCase().replace(/\b\w/g, (c) => c.toUpperCase());
}

// ---- State Normalization ----

const STATE_CODES: Record<string, string> = {
    alabama: 'AL', alaska: 'AK', arizona: 'AZ', arkansas: 'AR',
    california: 'CA', colorado: 'CO', connecticut: 'CT', delaware: 'DE',
    florida: 'FL', georgia: 'GA', hawaii: 'HI', idaho: 'ID',
    illinois: 'IL', indiana: 'IN', iowa: 'IA', kansas: 'KS',
    kentucky: 'KY', louisiana: 'LA', maine: 'ME', maryland: 'MD',
    massachusetts: 'MA', michigan: 'MI', minnesota: 'MN', mississippi: 'MS',
    missouri: 'MO', montana: 'MT', nebraska: 'NE', nevada: 'NV',
    'new hampshire': 'NH', 'new jersey': 'NJ', 'new mexico': 'NM', 'new york': 'NY',
    'north carolina': 'NC', 'north dakota': 'ND', ohio: 'OH', oklahoma: 'OK',
    oregon: 'OR', pennsylvania: 'PA', 'rhode island': 'RI', 'south carolina': 'SC',
    'south dakota': 'SD', tennessee: 'TN', texas: 'TX', utah: 'UT',
    vermont: 'VT', virginia: 'VA', washington: 'WA', 'west virginia': 'WV',
    wisconsin: 'WI', wyoming: 'WY', 'district of columbia': 'DC',
};

export const STATE_NAMES: Record<string, string> = Object.fromEntries(
    Object.entries(STATE_CODES).map(([name, code]) => [code, titleCase(name)])
);

function normalizeState(state?: string): string | null {
    if (!state) return null;
    const cleaned = state.trim().toLowerCase();
    if (cleaned.length === 2) {
        const upper = cleaned.toUpperCase();
        if (Object.values(STATE_CODES).includes(upper)) return upper;
    }
    return STATE_CODES[cleaned] || null;
}

function normalizePhone(phone?: string): string | null {
    if (!phone) return null;
    const digits = phone.replace(/\D/g, '');
    if (digits.length === 10) return `(${digits.slice(0, 3)}) ${digits.slice(3, 6)}-${digits.slice(6)}`;
    if (digits.length === 11 && digits[0] === '1') return `(${digits.slice(1, 4)}) ${digits.slice(4, 7)}-${digits.slice(7)}`;
    return phone.trim().slice(0, 20);
}

// ---- Dedup Hash ----

function generateDedupeHash(raw: RawListing): string {
    const parts = [
        raw.title?.toLowerCase().replace(/[^a-z0-9]/g, '').slice(0, 50) || '',
        parseCurrency(raw.askingPriceRaw)?.toString() || '',
        normalizeState(raw.state) || '',
        raw.city?.toLowerCase().replace(/[^a-z]/g, '').slice(0, 20) || '',
    ].join('|');
    return createHash('sha256').update(parts).digest('hex').slice(0, 16);
}

// ---- Industry Mapping ----

function mapIndustry(rawIndustry?: string, title?: string, desc?: string): string | null {
    if (!rawIndustry && !title) return null;
    const text = `${rawIndustry || ''} ${title || ''}`.toLowerCase();

    if (text.match(/restaurant|food|café|cafe|pizza|bar |grill|bakery|catering|coffee|deli/)) return 'Restaurants & Food';
    if (text.match(/saas|software|app |tech|platform|ai |machine learning|cloud|subscription/)) return 'Technology & SaaS';
    if (text.match(/ecommerce|e-commerce|shopify|amazon|fba|online store|dropship|marketplace/)) return 'E-commerce';
    if (text.match(/retail|store|shop(?!ify)|boutique|clothing|apparel|gift/)) return 'Retail';
    if (text.match(/construct|contractor|plumb|electric|hvac|roof|pav|build|renovation/)) return 'Construction';
    if (text.match(/health|medical|dental|clinic|pharmacy|chiro|therapy|physio|veterinar/)) return 'Healthcare & Medical';
    if (text.match(/manufactur|fabricat|produc|machining|assembly|industrial/)) return 'Manufacturing';
    if (text.match(/transport|truck|freight|logistic|moving|courier|delivery/)) return 'Transportation';
    if (text.match(/gas station|car wash|auto.*detail|lube|fuel/)) return 'Gas Stations & Car Washes';
    if (text.match(/auto|mechanic|tire|collision|body shop|dealership/)) return 'Automotive';
    if (text.match(/clean|janitorial|maid|laundry|dry clean|pressure wash/)) return 'Cleaning & Maintenance';
    if (text.match(/salon|spa|barber|beauty|nail|massage/)) return 'Beauty & Personal Care';
    if (text.match(/franchise/)) return 'Franchise';
    if (text.match(/hotel|motel|hostel|resort|airbnb|travel|tourism/)) return 'Travel & Hospitality';
    if (text.match(/media|marketing|agency|pr |publishing|advertising|seo/)) return 'Media & Marketing';
    if (text.match(/education|tutoring|school|training|daycare|childcare/)) return 'Education';
    if (text.match(/wholesale|distribut|import|export|supply chain/)) return 'Wholesale & Distribution';
    if (text.match(/financial|insurance|accounting|tax|bookkeeping|mortgage/)) return 'Finance & Insurance';
    if (text.match(/farm|agriculture|crop|livestock|nursery|landscap/)) return 'Agriculture';
    if (text.match(/gym|fitness|sport|recreation|entertainment|event|studio/)) return 'Entertainment & Recreation';
    if (text.match(/consult|professional|law|legal|engineer|architect/)) return 'Professional Services';
    if (text.match(/service/)) return 'Services';
    return 'Other';
}

File: scraper/core/normalizer.py
"""Currency, state, and industry normalizers."""
from __future__ import annotations
import re
import unicodedata
from typing import Optional


# ─── Currency ──────────────────────────────────────────────────────────────────

_MULTIPLIERS = {
    "b": 1_000_000_000,
    "billion": 1_000_000_000,
    "m": 1_000_000,
    "million": 1_000_000,
    "k": 1_000,
    "thousand": 1_000,
}


def parse_currency(raw: Optional[str]) -> Optional[int]:
    """
    Parse a human-readable money string into cents.

    Examples:
        "$1.2M" → 120000000
        "$250,000" → 25000000000... wait, *cents*:
        "$1.2M" → 120_000_000  (1.2M dollars = 120M cents)
        "$250,000" → 25_000_000 cents
        "Not Disclosed" → None
    """
    if not raw:
        return None

    raw = raw.strip()
    # Remove currency symbols and common no-data strings
    lower = raw.lower()
    if any(kw in lower for kw in ("not disclosed", "undisclosed", "n/a", "contact",
                                   "call", "inquire", "negotiable", "tbd", "--", "—")):
        return None

    # Normalize unicode
    raw = unicodedata.normalize("NFKD", raw)
    # Strip non-numeric prefix/suffix except digits, dots, commas, and letters
    cleaned = re.sub(r"[$€£¥₹,\s]", "", raw).lower()

    # Check for multiplier suffix
    multiplier = 1
    for suffix, mult in sorted(_MULTIPLIERS.items(), key=lambda x: -len(x[0])):
        if cleaned.endswith(suffix):
            multiplier = mult
            cleaned = cleaned[: -len(suffix)]
            break

    try:
        value_dollars = float(cleaned) * multiplier
        return int(value_dollars * 100)  # store as cents
    except (ValueError, TypeError):
        return None


def format_currency(cents: Optional[int]) -> Optional[str]:
    """Convert cents back to a human-readable string."""
    if cents is None:
        return None
    dollars = cents / 100
    if dollars >= 1_000_000_000:
        return f"${dollars / 1_000_000_000:.1f}B"
    if dollars >= 1_000_000:
        return f"${dollars / 1_000_000:.1f}M"
    if dollars >= 1_000:
        return f"${dollars / 1_000:.0f}K"
    return f"${dollars:,.0f}"


# ─── States ────────────────────────────────────────────────────────────────────

_STATE_MAP: dict[str, str] = {
    "alabama": "AL", "alaska": "AK", "arizona": "AZ", "arkansas": "AR",
    "california": "CA", "colorado": "CO", "connecticut": "CT", "delaware": "DE",
    "florida": "FL", "georgia": "GA", "hawaii": "HI", "idaho": "ID",
    "illinois": "IL", "indiana": "IN", "iowa": "IA", "kansas": "KS",
    "kentucky": "KY", "louisiana": "LA", "maine": "ME", "maryland": "MD",
    "massachusetts": "MA", "michigan": "MI", "minnesota": "MN", "mississippi": "MS",
    "missouri": "MO", "montana": "MT", "nebraska": "NE", "nevada": "NV",
    "new hampshire": "NH", "new jersey": "NJ", "new mexico": "NM", "new york": "NY",
    "north carolina": "NC", "north dakota": "ND", "ohio": "OH", "oklahoma": "OK",
    "oregon": "OR", "pennsylvania": "PA", "rhode island": "RI", "south carolina": "SC",
    "south dakota": "SD", "tennessee": "TN", "texas": "TX", "utah": "UT",
    "vermont": "VT", "virginia": "VA", "washington": "WA", "west virginia": "WV",
    "wisconsin": "WI", "wyoming": "WY", "district of columbia": "DC",
}

# Include both lower-case full names AND uppercase codes
_STATE_CODES: set[str] = set(_STATE_MAP.values())


def normalize_state(raw: Optional[str]) -> Optional[str]:
    if not raw:
        return None
    raw = raw.strip()
    if raw.upper() in _STATE_CODES:
        return raw.upper()
    return _STATE_MAP.get(raw.lower().strip())


# ─── Industry ──────────────────────────────────────────────────────────────────

_INDUSTRY_KEYWORDS: dict[str, list[str]] = {
    "Technology & SaaS": ["saas", "software", "tech", "app", "digital", "website", "ecommerce", "e-commerce", "ai", "fintech"],
    "Restaurants & Food": ["restaurant", "food", "bakery", "cafe", "bar", "catering", "pizza", "deli", "brewery"],
    "Retail": ["retail", "store", "shop", "boutique", "clothing", "apparel", "gift"],
    "Healthcare": ["medical", "dental", "pharmacy", "healthcare", "health", "clinic", "therapy", "care"],
    "Home Services": ["plumbing", "hvac", "landscaping", "cleaning", "roofing", "construction", "remodeling", "handyman", "pest"],
    "B2B Services": ["staffing", "accounting", "logistics", "distribution", "manufacturing", "wholesale", "consulting", "b2b"],
    "Education": ["education", "school", "tutoring", "daycare", "childcare", "learning"],
    "Automotive": ["auto", "car", "vehicle", "mechanic", "detailing", "repair"],
    "Fitness & Wellness": ["gym", "fitness", "yoga", "salon", "spa", "beauty", "wellness"],
    "Franchise": ["franchise"],
}


def normalize_industry(raw: Optional[str]) -> Optional[str]:
    if not raw:
        return None
    lower = raw.lower()
    for industry, keywords in _INDUSTRY_KEYWORDS.items():
        if any(kw in lower for kw in keywords):
            return industry
    # Return title-cased original if no match
    return raw.strip().title()[:100]


def clean_text(text: Optional[str], max_len: int = 5000) -> Optional[str]:
    if not text:
        return None
    text = text.strip()
    # Collapse whitespace
    text = re.sub(r"\s{2,}", " ", text)
    return text[:max_len] if text else None
